rasa/utils/tensorflow/model_data.py

```diff
@@ -31,14 +31,22 @@
 #   "numpy array containing dense features for every training example",
 #   "numpy array containing sparse features for every training example"
 # ]}
-Data = Dict[Text, Dict[Text, List[np.ndarray]]]
+# In case of e2e data, we also deal with 4d tensors, e.g.
+# "action_text" -> { "sequence": [
+#   [["numpy array containing dense features for every training example"]],
+#   [["numpy array containing sparse features for every training example"]]
+# ]}
+Features_4D = List[np.ndarray]
+Features_3D = np.ndarray
```

Comment:
hm, so list in 4D correspond to batch dim?, however in 3D it's first dim of ndarray.
I'm afraid it's prone to confusion after some time. Shall we create a special structure for Features, that would be basically ndarray, but contain all the info: whether it is sparse, whether it is 4D, which dimensions are what etc?

Reply:
and it seems to be if we keep the outer `ndarray`, it'll simplify things in the code. Can we subclass `ndarray`?

Reply:
why would you subclass `ndarray`?

I was first thinking of keeping both as numpy array with different dimensionalities, but then you would need to pad the data event before adding it to `RasaModelData` when creating the 4D features, which would have complicated the model data creation in TED. So, what do you suggest as a structure?

Reply:
but the ndarray in 3D features also contain non padded data, basically serving as a list

Reply:
> why would you subclass ndarray? what do you suggest as a structure?

smth that maintain the slicing of ndarray (needed for operating on batch dimension), but also containing the info about sparse, 4D, etc. to simplify our checks

Reply:
basically we're already using `ndarray` as ragged array. Because we're padding it only at last moment before creating tf.tensors

Reply:
I think we can even use ragged tensors, avoiding need for manual padding https://www.tensorflow.org/guide/ragged_tensor

Reply:
So, you are saying 4D should become np.array(np.array(np.array)) and 3D np.array(np.array)? And then we wrap it into some kind of structure so that we know if it is sparse, 4D, etc. So basically this object would already contain the information around FeatureSignature. Can try that.

I guess ragged tensors does not work for sparse, does it?

Reply:
ok, will try something, will ping you once it is ready for review

Reply:
I meant ragged tensors for dense

Reply:
>  3D np.array(np.array)

I believe we already have it

Reply:
yes, we do, just wanted to clarify